src/test/java/com/atlassian/db/replica/it/DualConnectionIT.java

```diff
@@ -43,45 +47,48 @@ public void shouldPreserveAutoCommitModeWhileSwitchingFromReplicaToMain() throws
     }
 
     @Test
-    public void shouldRunNextValOnMainDatabase() throws SQLException {
+    public void shouldPreserveReadOnlyModeWhileSwitchingFromReplicaToMain() throws SQLException {
         try (PostgresConnectionProvider connectionProvider = new PostgresConnectionProvider()) {
-            createTestSequence(connectionProvider);
-
+            final WaitingReplicaConsistency consistency = new WaitingReplicaConsistency(new LsnReplicaConsistency());
+            createTable(DualConnection.builder(connectionProvider, consistency).build());
             final Connection connection = DualConnection.builder(
                 connectionProvider,
-                CircularConsistency.permanentConsistency().build()
+                consistency
             ).build();
 
+            connection.setAutoCommit(false);
+            connection.setReadOnly(true);
+            final PreparedStatement preparedStatement = connection.prepareStatement("INSERT INTO foo(bar) VALUES(?);");
+            preparedStatement.setString(1, "test");
+
+            final Throwable throwable = catchThrowable(preparedStatement::executeUpdate);
+
+            Assertions.assertThat(connection.isReadOnly()).isTrue();
+            Assertions.assertThat(throwable).hasMessage("ERROR: cannot execute INSERT in a read-only transaction");
+        }
+    }
+
+    @Test
+    public void shouldRunNextValOnMainDatabase() throws SQLException {
+        try (PostgresConnectionProvider connectionProvider = new PostgresConnectionProvider()) {
+            final WaitingReplicaConsistency consistency = new WaitingReplicaConsistency(new LsnReplicaConsistency());
+            createTestSequence(DualConnection.builder(connectionProvider, consistency).build());
+            final Connection connection = DualConnection.builder(connectionProvider, consistency).build();
+
             connection.prepareStatement("SELECT nextval('test_sequence');").executeQuery();
         }
     }
 
-    private void createTestSequence(PostgresConnectionProvider connectionProvider) throws SQLException {
-        final Connection mainConnection = connectionProvider.getMainConnection();
-        try (final Statement mainStatement = mainConnection.createStatement()) {
+    private void createTestSequence(Connection connection) throws SQLException {
+        try (final Statement mainStatement = connection.createStatement()) {
             mainStatement.execute("CREATE SEQUENCE test_sequence;");
-            final LsnReplicaConsistency consistency = new LsnReplicaConsistency();
-            waitForReplicaConsistency(connectionProvider, mainConnection, consistency);
         }
     }
 
-    private void waitForReplicaConsistency(
```

Comment:
Extracted to `WaitingReplicaConsistency ` 👍

Reply:
Maybe not exactly this implementation, but something similar could work in production code for some not critical, but SQL heavy code paths (generating reports, backups ...).